presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -549,6 +569,34 @@ public void removePreparedStatement(String key)
         deallocatedPreparedStatements.add(key);
     }
 
+    public void addSessionFunction(BasicSignature signature, SqlInvokedFunction function)
+    {
+        requireNonNull(signature, "signature is null");
+        requireNonNull(function, "function is null");
+
+        String serializedSignature = BasicSignatureSerDe.serialize(signature);
+        String serializedFunction = SqlInvokedFunctionSerDe.serialize(function);
+
+        if (session.getSessionFunctions().containsKey(serializedSignature) || addedSessionFunctions.putIfAbsent(serializedSignature, serializedFunction) != null) {
+            throw new PrestoException(ALREADY_EXISTS, String.format("Session function %s has already been defined", signature));
+        }
+    }
+
+    public void removeSessionFunction(BasicSignature signature, boolean suppressNotFoundException)
```

Comment:
What you can do is to return a boolean to indicate success / failure. Throw on the caller side if needed.

Reply:
I think it is a cleaner API to have QueryStateMachine throw, and this way is consistent with how prepared statements are handled in this class.